File: backend/services/offer_service.py
"""
Main offer generation orchestration service.
Coordinates scoring, pricing, LLM copy generation, and offer persistence.
"""
from typing import Dict, List, Any, Optional, Tuple
import json
from datetime import datetime, timedelta
import sqlite3
import uuid
import os

from services.scoring_service import compute_score, filter_eligible_candidates
from services.pricing_service import calculate_offer_pricing, validate_pricing
from services.rag_service import generate_marketing_copy, generate_property_diffs, generate_full_offer_preview_copy
from database import get_db


def generate_headline_and_summary(
    candidate_prop: Dict[str, Any],
    diffs: List[str],
    booking: Dict[str, Any]
) -> Tuple[str, str]:
    """
    Generate headline and summary for upgrade option.
    
    Returns:
        (headline, summary) tuple
    """
    # Simple headline generation
    prop_name = candidate_prop["name"]
    
    # Parse metadata
    metadata = candidate_prop.get("metadata", {})
    if isinstance(metadata, str):
        metadata = json.loads(metadata) if metadata else {}
    
    # Create headline based on property features
    if "pool" in candidate_prop.get("amenities", []):
        headline = "Upgrade to villa with private pool"
    elif candidate_prop["beds"] >= 3:
        headline = "More space for the whole family"
    elif "beachfront" in metadata.get("beach_distance", ""):
        headline = "Beachfront luxury upgrade"
    else:
        headline = f"Upgrade to {prop_name}"
    
    # Create summary from diffs
    if len(diffs) >= 2:
        summary = f"{diffs[0]}, {diffs[1].lower()}"
        if len(diffs) >= 3:
            summary += f", {diffs[2].lower()}"
        summary += "."
    elif diffs:
        summary = diffs[0] + "."
    else:
        summary = f"Enhanced property in {candidate_prop['location']}."
    
    return headline, summary


def get_all_properties(conn) -> List[Dict[str, Any]]:
    """Fetch all properties from database."""
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM properties")
    rows = cursor.fetchall()
    
    properties = []
    for row in rows:
        prop = dict(row)
        properties.append(prop)
    
    return properties


def get_booking(conn, booking_id: str) -> Optional[Dict[str, Any]]:
    """Fetch booking by ID."""
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM bookings WHERE id = ?", (booking_id,))
    row = cursor.fetchone()
    
    if row:
        return dict(row)
    return None


def get_property(conn, prop_id: int) -> Optional[Dict[str, Any]]:
    """Fetch property by ID."""
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM properties WHERE id = ?", (prop_id,))
    row = cursor.fetchone()
    
    if row:
        return dict(row)
    return None


def get_host_settings(conn, host_id: str) -> Dict[str, Any]:
    """Fetch host settings."""
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM host_settings WHERE host_id = ?", (host_id,))
    row = cursor.fetchone()
    
    if row:
        settings = dict(row)
        # Parse JSON fields
        if settings.get("blocked_prop_ids"):
            try:
                settings["blocked_prop_ids"] = json.loads(settings["blocked_prop_ids"])
            except:
                settings["blocked_prop_ids"] = []
        return settings
    
    # Return defaults if not found
    return {
        "host_id": host_id,
        "max_discount_pct": 0.40,
        "min_revenue_lift_eur_per_night": 30.0,
        "max_adr_multiplier": 2.5,
        "offer_validity_hours": 48,
        "use_openai_for_copy": False,
        "local_llm_model": "gemma3:latest"
    }


def generate_offer(booking_id: str) -> Optional[str]:
    """
    Main offer generation function.
    """
    print(f"\n=== Starting offer generation for booking {booking_id} ===")
    offer_id = str(uuid.uuid4())
    with get_db() as conn:
        # 1. Load booking and original property
        print(f"[1/7] Loading booking data...")
        booking = get_booking(conn, booking_id)
        if not booking:
            print(f"❌ Booking {booking_id} not found")
            return None
        print(f"✓ Booking loaded: {booking['guest_name']} at {booking.get('prop_id')}")
            
        # Clear any existing offers for this booking (for demo re-runnability)
        print(f"[2/7] Clearing existing offers...")
        conn.execute("DELETE FROM offers WHERE booking_id = ?", (booking_id,))
        conn.commit()
        
        original_prop = get_property(conn, booking["prop_id"])
        if not original_prop:
            print(f"❌ Property {booking['prop_id']} not found")
            return None
        print(f"✓ Original property: {original_prop['name']}")
        
        # 2. Load host settings
        print(f"[3/7] Loading host settings...")
        host_id = booking.get("host_id", "demo_host_001")
        host_settings = get_host_settings(conn, host_id)
        print(f"✓ Host settings loaded (use_openai={host_settings.get('use_openai_for_copy', False)})")
        
        # 3. Find eligible candidates
        print(f"[4/7] Finding eligible upgrade candidates...")
        all_properties = get_all_properties(conn)
        candidates = filter_eligible_candidates(
            all_properties,
            original_prop,
            booking,
            host_settings
        )
        
        if not candidates:
            print("❌ No eligible upgrade candidates found")
            return None
        print(f"✓ Found {len(candidates)} eligible candidates")
        
        # 4. Score each candidate and calculate pricing
        print(f"[5/7] Scoring candidates and generating AI copy...")
        scored_options = []
        for candidate in candidates:
            # Compute viability score
            score = compute_score(original_prop, candidate, booking)
            
            # Calculate pricing
            pricing = calculate_offer_pricing(
                from_adr=booking["base_nightly_rate"],
                to_adr=candidate["list_nightly_rate"],
                nights=booking["nights"],
                discount_pct=host_settings.get("max_discount_pct", 0.40),
                from_total=booking["total_paid"]
            )
            
            # Validate pricing against guardrails
            if not validate_pricing(pricing, host_settings):
                continue
            
            # Generate diffs
            diffs = generate_property_diffs(original_prop, candidate)
            
            # Generate fallback headline/summary
            headline, summary = generate_headline_and_summary(candidate, diffs, booking)
            
            # Generate Full AI Copy Package
            ai_data = generate_full_offer_preview_copy(
                original_prop,
                candidate,
                pricing,
                {
                    "guest_name": booking["guest_name"],
                    "adults": booking.get("adults", 2),
                    "children": booking.get("children", 0),
                    "local_model": host_settings.get("local_llm_model", "gemma3:latest")
                },
                use_openai=bool(host_settings.get("use_openai_for_copy", False))
            )
            
            # Parse images
            images = candidate.get("images", [])
            if isinstance(images, str):
                images = json.loads(images) if images else []
            
            option = {
                "ranking": 0,  # Will be set after sorting
                "prop_id": candidate["id"],
                "prop_name": candidate["name"],
                "viability_score": score,
                "pricing": pricing,
                "diffs": ai_data.get("diff_bullets", diffs), # Use high-quality bullets
                "headline": ai_data.get("landing_hero", headline),
                "summary": ai_data.get("landing_summary", summary),
                "ai_copy": ai_data, # Store the whole package
                "images": images,
                "availability": {
                    "available": True,
                    "reason": None
                }
            }
            
            scored_options.append(option)
        
        # 5. Sort by score and take top 3
        print(f"[6/7] Selecting top 3 options...")
        scored_options.sort(key=lambda x: x["viability_score"], reverse=True)
        top3 = scored_options[:3]
        
        if not top3:
            print("❌ No valid upgrade options after filtering")
            return None
        print(f"✓ Top 3 selected (scores: {[opt['viability_score'] for opt in top3]})")
        
        # Set rankings
        for idx, option in enumerate(top3):
            option["ranking"] = idx + 1
        
        # 6. Store offer
        print(f"[7/7] Storing offer in database...")
        now_dt = datetime.utcnow()
        validity_hours = host_settings.get("offer_validity_hours", 48)
        expires_dt = now_dt + timedelta(hours=validity_hours)
        
        # Hard cap at arrival date (e.g. 10am on arrival day)
        try:
            arrival_dt = datetime.strptime(booking["arrival_date"], "%Y-%m-%d")
            # Set to 10:00 AM on arrival day
            arrival_limit = arrival_dt.replace(hour=10, minute=0, second=0)
            
            if expires_dt > arrival_limit:
                expires_dt = arrival_limit
        except Exception as e:
            print(f"⚠️  Error parsing arrival date for expiration: {e}")
            
        now = now_dt.isoformat() + "Z"
        expires_at = expires_dt.isoformat() + "Z"
        
        # Prepare email data from the top option
        top_option = top3[0]
        email_subject = top_option.get("ai_copy", {}).get("subject", f"{booking['guest_name']}, upgrade your stay?")
        email_body = top_option.get("ai_copy", {}).get("email_html", "")
        
        # Inject Offer URL
        frontend_url = os.getenv("NEXT_PUBLIC_FRONTEND_URL") or "http://localhost:3030"
        frontend_url = frontend_url.rstrip('/')
        offer_url = f"{frontend_url}/offer/{offer_id}"
        email_body = email_body.replace("{OFFER_URL}", offer_url)

        cursor = conn.cursor()
        cursor.execute("""
        INSERT INTO offers 
        (id, booking_id, status, top3, email_subject, email_body_html, expires_at, created_at)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?)
        """, (
            offer_id,
            booking_id,
            "active",
            json.dumps(top3),
            email_subject,
            email_body,
            expires_at,
            now
        ))
        
        conn.commit()
        
        print(f"✅ Generated offer {offer_id} with {len(top3)} options")
        print(f"=== Offer generation complete ===")
        return offer_id
def find_overlapping_bookings(conn, prop_id: str, arrival: str, departure: str) -> List[Dict[str, Any]]:
    """
    Find bookings that overlap with the cancelled dates but are in different properties.
    """
    cursor = conn.cursor()
    # Logic: Search for bookings that stay during the same window but could move to the vacated prop
    # We exclude the same property (obviously) and look for lower-tier properties
    cursor.execute("""
        SELECT b.*, p.list_nightly_rate
        FROM bookings b
        JOIN properties p ON b.prop_id = p.id
        WHERE b.arrival_date >= ? AND b.departure_date <= ?
        AND b.prop_id != ?
    """, (arrival, departure, prop_id))
    
    return [dict(row) for row in cursor.fetchall()]

def handle_cancellation(cancelled_booking_id: str) -> List[str]:
    """
    Orchestrates the response to a cancellation.
    Returns a list of generated offer IDs.
    """
    generated_offers = []
    with get_db() as conn:
        # 1. Get the cancelled booking details
        cursor = conn.cursor()
        cursor.execute("SELECT * FROM bookings WHERE id = ?", (cancelled_booking_id,))
        cancelled = cursor.fetchone()
        if not cancelled:
            return []
            
        cancelled = dict(cancelled)
        
        # 2. Find overlapping candidates
        candidates = find_overlapping_bookings(
            conn, 
            cancelled["prop_id"], 
            cancelled["arrival_date"], 
            cancelled["departure_date"]
        )
        
        if not candidates:
            print(f"No candidates for cancellation recovery of booking {cancelled_booking_id}")
            return []
            
        # 3. Target the top 3 best guests (simplification for demo)
        # In a real system, we might only target the 'biggest' upgrade or send multiple
        for guest_booking in candidates[:3]:
            offer_id = generate_offer(guest_booking["id"])
            if offer_id:
                generated_offers.append(offer_id)
                # In demo mode, we also trigger the test email immediately if configured
                from services.email_service import send_test_email
                # Fetch the offer we just created to get the subject/body
                cursor.execute("SELECT * FROM offers WHERE id = ?", (offer_id,))
                offer_row = cursor.fetchone()
                if offer_row:
                    offer_data = dict(offer_row)
                    send_test_email(
                        to_email=guest_booking["guest_email"], 
                        subject=offer_data["email_subject"],
                        html_content=offer_data["email_body_html"]
                    )
                    
    return generated_offers

File: frontend/lib/services/offerService.ts
import { db } from '../db';
import { Booking, Property, Offer, UpgradeOption, PricingDetails, HostSettings } from '../../types';
import { v4 as uuidv4 } from 'uuid';
import OpenAI from 'openai';
import * as sgMail from '@sendgrid/mail';
import React from 'react';

import { EasyRenderer } from '../../components/EasyRenderer';
import { getOfferContext, defaultTemplate, getAcceptanceContext, defaultAcceptanceTemplate } from '../templateUtils';

const openai = new OpenAI({
    apiKey: process.env.OPENAI_API_KEY,
});

if (process.env.SENDGRID_API_KEY) {
    sgMail.setApiKey(process.env.SENDGRID_API_KEY);
}

export async function generateOffer(bookingId: string, sessionId?: string): Promise<string | null> {
    console.log(`\n=== Starting TS offer generation for booking ${bookingId} (Session: ${sessionId || 'None'}) ===`);

    const booking = await db.getBooking(bookingId, sessionId);
    if (!booking) {
        console.error(`❌ Booking ${bookingId} not found`);
        return null;
    }

    const originalProp = await db.getProperty(booking.prop_id);
    if (!originalProp) {
        console.error(`❌ Property ${booking.prop_id} not found`);
        return null;
    }

    const hostSettings = await db.getHostSettings(booking.host_id || 'demo_host_001');
    const useOpenAI = process.env.NEXT_PUBLIC_USE_OPENAI === 'true' || hostSettings?.use_openai_for_copy;

    // 1. Initial Filtering & Scoring
    const allProperties = await db.getAllProperties();
    const candidates = allProperties.filter(p => p.id !== booking.prop_id);

    console.log(`Analyzing ${candidates.length} candidates for booking ${bookingId}`);

    const scoredOptions: UpgradeOption[] = [];

    for (const candidate of candidates) {
        // Basic eligibility check
        if (candidate.beds < originalProp.beds) continue; // Don't offer smaller places

        const score = computeScore(originalProp, candidate, booking);

        const pricing = calculateOfferPricing(
            booking.base_nightly_rate,
            candidate.list_nightly_rate,
            booking.nights,
            hostSettings?.max_discount_pct ?? 0.45,
            booking.total_paid,
            hostSettings?.min_revenue_lift_eur_per_night ?? 15.00
        );

        const diffs = generatePropertyDiffs(originalProp, candidate);
        const candidateImages = parseImages(candidate.images);
        const images = candidateImages.length > 0 ? candidateImages : [`/properties/${candidate.id}.png`];

        scoredOptions.push({
            ranking: 0,
            prop_id: candidate.id,
            prop_name: candidate.name,
            viability_score: score,
            pricing,
            diffs,
            headline: `Upgrade to ${candidate.name}`,
            summary: `Experience more space and better amenities.`,
            images: images,
            ai_copy: null,
            amenities: parseAmenities(candidate.amenities),
            metadata: candidate.metadata || {},
            availability: { status: 'available' }
        });
    }

    // Sort by score and take top 3
    const topCandidates = scoredOptions
        .sort((a, b) => b.viability_score - a.viability_score)
        .slice(0, 3);

    if (topCandidates.length === 0) {
        console.warn(`⚠️ No suitable upgrade options found for booking ${bookingId}`);
        return null;
    }

    // 2. Generate AI Copy only for the #1 pick
    console.log(`Generating AI copy for the best candidate: ${topCandidates[0].prop_name}...`);
    const top3 = await Promise.all(topCandidates.map(async (opt, idx) => {
        let aiCopy = null;
        // Only generate AI copy for the top one to be fast and reliable
        if (useOpenAI && idx === 0) {
            try {
                const candidate = allProperties.find(p => p.id === opt.prop_id)!;
                aiCopy = await generateAICopy(originalProp, candidate, opt.pricing, booking, opt.diffs, hostSettings);
                console.log(`✅ AI Copy generated for ${opt.prop_name}`);
            } catch (err) {
                console.error(`❌ AI Generation failed for ${opt.prop_name}:`, err);
            }
        }
        return {
            ...opt,
            ranking: idx + 1,
            ai_copy: aiCopy,
            headline: aiCopy?.landing_hero || opt.headline,
            summary: aiCopy?.landing_summary || opt.summary
        };
    }));

    const offerId = uuidv4();
    const expiresAt = new Date();
    expiresAt.setHours(expiresAt.getHours() + (hostSettings?.offer_validity_hours || 48));

    const bestOption = top3[0];
    const baseUrl = (process.env.NEXT_PUBLIC_FRONTEND_URL || 'https://uprez.dpitcock.dev').replace(/\/$/, '');
    const offerUrl = `${baseUrl}/offer/${offerId}`;
    const companyName = hostSettings?.pm_company_name || hostSettings?.host_name || 'Your Host';

    // Ensure we have a valid image URL for the email
    const mainImage = bestOption.images?.[0];
    const imageUrl = mainImage
        ? (mainImage.startsWith('http') ? mainImage : `${baseUrl}${mainImage.startsWith('/') ? '' : '/'}${mainImage}`)
        : `${baseUrl}/properties/${bestOption.prop_id}.png`;

    // Calculate per-night upgrade fee
    const upgradePerNight = Math.round(bestOption.pricing.revenue_lift / bestOption.pricing.nights);

    const expiresStr = `${new Date(expiresAt).toLocaleDateString()} at ${new Date(expiresAt).toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' })}`;

    const context = getOfferContext({ id: offerId, expires_at: expiresAt.toISOString() } as Offer, booking, top3);
    // @ts-ignore - Dynamic require to avoid build errors in Next.js App Router
    const { renderToStaticMarkup } = require('react-dom/server');
    const emailHtml = renderToStaticMarkup(
        React.createElement(EasyRenderer, {
            templateJson: defaultTemplate,
            mode: 'email',
            data: context
        })
    );

    const offer: Partial<Offer> = {
        id: offerId,
        booking_id: bookingId,
        status: 'active',
        top3,
        expires_at: expiresAt.toISOString(),
        email_subject: bestOption.ai_copy?.subject || `Exclusive Upgrade Opportunity: ${bestOption.prop_name}`,
        email_body_html: emailHtml,
        created_at: new Date().toISOString()
    };

    console.log(`💾 Saving offer ${offerId} to database...`);
    await db.saveOffer(offer, sessionId);

    // Trigger email
    const recipient = process.env.CONTACT_EMAIL || booking.guest_email;
    console.log(`📧 Attempting to send email to ${recipient}...`);
    await sendOfferEmail(offer as Offer, booking);

    return offerId;
}

export async function generateOptionCopy(offerId: string, propId: string, sessionId?: string): Promise<any> {
    console.log(`\n=== Generating on-demand copy for offer ${offerId}, property ${propId} ===`);

    const offer = await db.getOffer(offerId); // Offer ID is unique UUID, so no session scoping needed for lookup?
    // Wait, getOffer does strictly ID lookup.
    // If offer ID is UUID, it's globally unique.
    if (!offer) throw new Error('Offer not found');

    const booking = await db.getBooking(offer.booking_id, sessionId);
    if (!booking) throw new Error('Booking not found');

    const originalProp = await db.getProperty(booking.prop_id);
    if (!originalProp) throw new Error('Original property not found');

    const top3 = offer.top3 as UpgradeOption[];
    const optionIndex = top3.findIndex(o => o.prop_id === propId);
    if (optionIndex === -1) throw new Error('Option not found in offer');

    const option = top3[optionIndex];
    const candidateProp = await db.getProperty(propId);
    if (!candidateProp) throw new Error('Candidate property not found');

    const hostSettings = await db.getHostSettings(booking.host_id || 'demo_host_001');

    // Generate AI copy with host settings context if available
    const aiCopy = await generateAICopy(originalProp, candidateProp, option.pricing, booking, option.diffs, hostSettings);

    // Update the option in the list
    top3[optionIndex] = {
        ...option,
        ai_copy: aiCopy,
        headline: aiCopy.landing_hero || option.headline,
        summary: aiCopy.landing_summary || option.summary
    };

    // Update the offer in DB
    await db.saveOffer({
        ...offer,
        top3
    }, sessionId);

    console.log(`✅ On-demand copy generated and saved for ${propId}`);
    return aiCopy;
}

export async function acceptOffer(offerId: string, propId: string, sessionId?: string): Promise<any> {
    console.log(`\n=== Accepting offer ${offerId}, property ${propId} ===`);

    const offer = await db.getOffer(offerId);
    if (!offer) throw new Error('Offer not found');
    if (offer.status !== 'active') throw new Error(`Offer is already ${offer.status}`);

    const booking = await db.getBooking(offer.booking_id, sessionId);
    if (!booking) throw new Error('Booking not found');

    const hostSettings = await db.getHostSettings(booking.host_id || 'demo_host_001');

    const top3 = offer.top3 as UpgradeOption[];
    const option = top3.find(o => o.prop_id === propId);
    if (!option) throw new Error('Selected property not found in offer options');

    // 1. Update Offer Status
    await db.updateOfferStatus(offerId, 'accepted', propId);

    // 2. Update Booking
    await db.updateBooking(offer.booking_id, {
        prop_id: propId,
        status: 'upgraded',
        upgraded_from_prop_id: booking.prop_id,
        base_nightly_rate: option.pricing.offer_adr,
        total_paid: option.pricing.offer_total,
        upgrade_at: new Date().toISOString()
    });

    // 3. Expire all other active offers for the same property
    // This prevents double-booking when multiple guests are offered the same upgrade
    try {
        const { neon } = require('@neondatabase/serverless');
        const sql = neon(process.env.POSTGRES_URL!);

        const result = await sql`
            UPDATE offers 
            SET status = 'expired', updated_at = ${new Date().toISOString()}
            WHERE status = 'active' 
            AND id != ${offerId}
            AND top3::text LIKE ${`%"prop_id":"${propId}"%`}
        `;

        const expiredCount = result.length || 0;
        if (expiredCount > 0) {
            console.log(`🚫 Expired ${expiredCount} other active offer(s) for property ${propId}`);
        }
    } catch (err) {
        console.error('⚠️ Failed to expire competing offers:', err);
        // Don't fail the acceptance if this step fails - log and continue
    }

    // 4. Send Confirmation Email
    try {
        await sendAcceptanceEmail(booking, option, hostSettings);
    } catch (err) {
        console.error('⚠️ Failed to send acceptance email:', err);
    }

    console.log(`✅ Offer ${offerId} accepted. Booking ${offer.booking_id} updated to property ${propId}.`);

    return {
        success: true,
        message: 'Upgrade successful',
        booking_id: offer.booking_id
    };
}

export function computeScore(orig: Property, cand: Property, booking: Booking): number {
    let score = 5.0;

    // Better amenities
    if (cand.beds > orig.beds) score += 2.0;
    if (cand.baths > orig.baths) score += 1.0;

    // Location context (simulated)
    if (cand.location === orig.location) score += 1.0;

    // Price delta context
    const priceRatio = cand.list_nightly_rate / orig.list_nightly_rate;
    if (priceRatio > 1.1 && priceRatio < 2.5) score += 1.0;

    return Math.min(10, score);
}

export function calculateOfferPricing(
    fromAdr: number,
    toAdr: number,
    nights: number,
    maxDiscountPct: number,
    fromTotal: number,
    minLiftPerNight: number = 15
): PricingDetails {
    const stayTotalAtListRate = toAdr * nights;
    const currentTotal = fromTotal;

    // The maximum possible lift (rack rate difference)
    const rawTotalDifference = stayTotalAtListRate - currentTotal;
    const rawLiftPerNight = rawTotalDifference / nights;

    // Apply the configured discount
    const discountedLiftPerNight = rawLiftPerNight * (1 - maxDiscountPct);

    // Guardrail: Ensure we don't go below minLiftPerNight, 
    // but also don't exceed the raw difference (don't charge more than rack rate)
    const finalLiftPerNight = Math.max(
        discountedLiftPerNight,
        Math.min(rawLiftPerNight, minLiftPerNight)
    );

    const upgradeFeeTotal = finalLiftPerNight * nights;
    const offerTotal = currentTotal + upgradeFeeTotal;
    const offerAdr = offerTotal / nights;

    // Calculate effective discount percentage for the UI
    const totalDiscountAmount = Math.max(0, stayTotalAtListRate - offerTotal);
    const effectiveDiscountPercent = stayTotalAtListRate > 0
        ? Math.round((totalDiscountAmount / stayTotalAtListRate) * 100)
        : 0;

    return {
        currency: 'EUR',
        from_adr: fromAdr,
        to_adr_list: toAdr,
        offer_adr: offerAdr,
        nights,
        from_total: fromTotal,
        offer_total: offerTotal,
        list_total: stayTotalAtListRate,
        discount_percent: effectiveDiscountPercent,
        discount_amount_total: totalDiscountAmount,
        revenue_lift: upgradeFeeTotal
    };
}

export function generatePropertyDiffs(orig: Property, cand: Property): string[] {
    const diffs: string[] = [];
    if (cand.beds > orig.beds) diffs.push(`${cand.beds - orig.beds} Extra Bedroom(s)`);
    if (cand.baths > orig.baths) diffs.push(`Additional Bathroom`);
    // Add more logic here as needed
    return diffs.slice(0, 3);
}

async function generateAICopy(orig: Property, cand: Property, pricing: PricingDetails, booking: Booking, diffs: string[], hostSettings?: HostSettings | null) {
    const companyName = hostSettings?.pm_company_name || hostSettings?.host_name || 'your host';
    const upgradePerNight = Math.round(pricing.revenue_lift / pricing.nights);

    const prompt = `Generate a luxury upgrade offer for ${booking.guest_name} from ${companyName}. 
    
    Original property: ${orig.name} at ${pricing.from_adr}€/night (Total paid: €${pricing.from_total})
    Upgrade property: ${cand.name} at ${pricing.to_adr_list}€/night list rate
    
    PRICING (IMPORTANT):
    - Upgrade fee: €${upgradePerNight}/night (€${pricing.revenue_lift} total for ${pricing.nights} nights)
    - This is the ADDITIONAL cost to upgrade, not the new total price
    - Display as: "Upgrade for just €${upgradePerNight}/night" or similar language that emphasizes the incremental cost
    
    Property differences: ${diffs.join(', ')}.
    
    Return JSON with the following structure:
    { 
      "subject": "Email subject line", 
      "email_title": "Primary greeting/title in the email (e.g. Great news, [Name]!)",
      "email_content": "A short, persuasive paragraph about why this upgrade is perfect for them.",
      "email_selling_points": ["Point 1", "Point 2", "Point 3"],
      "email_cta": "CTA Button text (e.g. VIEW MY UPGRADE)",
      "landing_hero": "Hero headline for the webpage", 
      "landing_summary": "Summary text for the webpage", 
      "diff_bullets": ["Bullet 1", "Bullet 2", "Bullet 3"] 
    }`;

    const response = await openai.chat.completions.create({
        model: "gpt-4o-mini",
        messages: [{ role: "system", content: "You are a luxury concierge content generator." }, { role: "user", content: prompt }],
        response_format: { type: "json_object" }
    });

    return JSON.parse(response.choices[0].message.content || '{}');
}

async function sendAcceptanceEmail(booking: Booking, option: UpgradeOption, hostSettings: HostSettings | null) {
    if (process.env.EMAIL_ENABLED !== 'true') return;

    if (!process.env.SENDGRID_API_KEY) {
        console.error('❌ SENDGRID_API_KEY is missing (Acceptance Email)');
        return;
    }

    const recipient = process.env.CONTACT_EMAIL || booking.guest_email;
    const companyName = hostSettings?.pm_company_name || hostSettings?.host_name || 'Your Host';

    const context = getAcceptanceContext(booking, option, hostSettings);
    // @ts-ignore
    const { renderToStaticMarkup } = require('react-dom/server');
    const emailHtml = renderToStaticMarkup(
        React.createElement(EasyRenderer, {
            templateJson: defaultAcceptanceTemplate,
            mode: 'email',
            data: context
        })
    );

    const msg = {
        to: recipient,
        from: process.env.SENDGRID_FROM_EMAIL || 'dpitcock.dev@gmail.com',
        subject: `Upgrade Confirmed: ${option.prop_name}`,
        text: `Hi ${booking.guest_name}, Great news! Your upgrade to ${option.prop_name} has been confirmed. Total Paid: €${option.pricing.offer_total}. Dates: ${new Date(booking.arrival_date).toLocaleDateString()} - ${new Date(booking.departure_date).toLocaleDateString()}. Best, ${companyName}`,
        html: emailHtml
    };

    try {
        await sgMail.send(msg);
        console.log(`✅ Acceptance email sent to ${recipient}`);
    } catch (error: any) {
        console.error('❌ SendGrid Error (Acceptance):', error.response?.body || error.message);
    }
}

async function sendOfferEmail(offer: Offer, booking: Booking) {
    if (process.env.EMAIL_ENABLED !== 'true') {
        console.log('ℹ️ Email sending is disabled (EMAIL_ENABLED != true)');
        return;
    }
    if (!process.env.SENDGRID_API_KEY) {
        console.error('❌ SENDGRID_API_KEY is missing');
        return;
    }

    const recipient = process.env.CONTACT_EMAIL || booking.guest_email;

    const msg = {
        to: recipient,
        from: process.env.SENDGRID_FROM_EMAIL || 'dpitcock.dev@gmail.com',
        subject: offer.email_subject || 'Your Upgrade Offer',
        html: offer.email_body_html || '<p>Error loading content.</p>',
    };

    try {
        await sgMail.send(msg);
        console.log(`✅ Email sent to ${recipient} (Provider: SendGrid)`);
    } catch (error: any) {
        console.error('❌ SendGrid Error:', error.response?.body || error.message);
    }
}

/**
 * Diagnostic helper to test email integration directly
 */
export async function sendTestEmail(to: string) {
    console.log(`🚀 Sending test email to ${to}...`);

    if (process.env.EMAIL_ENABLED !== 'true') {
        throw new Error('Email is disabled in configuration');
    }

    // @ts-ignore
    const { renderToStaticMarkup } = require('react-dom/server');
    const emailHtml = renderToStaticMarkup(
        React.createElement(EasyRenderer, {
            templateJson: {
                root: {
                    children: [
                        {
                            component: 'Hero',
                            props: {
                                headline: 'UpRez Integration Active',
                                subheadline: 'If you see this, your SendGrid configuration is correct.'
                            }
                        }
                    ]
                }
            },
            mode: 'email',
            data: {}
        })
    );

    const msg = {
        to,
        from: process.env.SENDGRID_FROM_EMAIL || 'dpitcock.dev@gmail.com',
        subject: 'UpRez Integration Test',
        html: emailHtml
    };

    return sgMail.send(msg);
}

function parseImages(imgs: any): string[] {
    if (Array.isArray(imgs)) return imgs;
    if (typeof imgs === 'string') {
        try { return JSON.parse(imgs); } catch { return []; }
    }
    return [];
}

function parseAmenities(am: any): string[] {
    if (Array.isArray(am)) return am;
    if (typeof am === 'string') {
        try { return JSON.parse(am); } catch { return am.split(',').map(s => s.trim()); }
    }
    return [];
}



